source/ada/lsp-ada_handlers.adb

```diff
@@ -1808,8 +1817,131 @@ package body LSP.Ada_Handlers is
+   begin
+      Self.Token_Id := Self.Token_Id + 1;
+      --  Generate an identifier that has little risk of collision with
+      --  other language servers, or other occurrences of this server.
+      --  (There is still a very small risk of collision with PID recyclings,
+      --  but the consequences are acceptable.)
+      return
+        (Is_Number => False,
+         String    => To_LSP_String
+           ("ada_ls-"
+            & Pid & "-" & Operation & "-"
+            & GNATCOLL.Utils.Image (Self.Token_Id, 1)));
+   end Get_Unique_Progress_Token;
+
+   -----------------
+   -- Index_Files --
+   -----------------
+
+   procedure Index_Files (Self : access Message_Handler) is
+      type Context_And_Files is record
+         Context : Context_Access;
+         Sources : File_Array_Access;
+      end record;
+      type Context_And_Files_Array is array
+        (Natural range <>) of Context_And_Files;
+
+      Arr   : Context_And_Files_Array (1 .. Natural (Self.Contexts.Length));
+
+      token : constant LSP.Types.LSP_Number_Or_String
+        := Self.Get_Unique_Progress_Token ("indexing");
+
+      procedure Emit_Progress_Begin;
+      procedure Emit_Progress_Report (Percent : Natural);
+      procedure Emit_Progress_End;
+      --  Emit a message to inform that the indexing has begun / is in
+      --  progress / has finished.
+
+      -------------------------
+      -- Emit_Progress_Begin --
+      -------------------------
+
+      procedure Emit_Progress_Begin is
+         P : LSP.Messages.Progress_Params (LSP.Messages.Progress_Begin);
+      begin
+         P.Begin_Param.token := token;
+         P.Begin_Param.value.title := LSP.Types.To_LSP_String ("Indexing");
+         P.Begin_Param.value.percentage := (Is_Set => True, Value => 0);
+         Self.Server.On_Progress (P);
+      end Emit_Progress_Begin;
+
+      --------------------------
+      -- Emit_Progress_Report --
+      --------------------------
+
+      procedure Emit_Progress_Report (Percent : Natural) is
+         P : LSP.Messages.Progress_Params (LSP.Messages.Progress_Report);
+      begin
+         P.Report_Param.token := token;
+         P.Report_Param.value.percentage := (Is_Set => True, Value => Percent);
+         Self.Server.On_Progress (P);
+      end Emit_Progress_Report;
+
+      -----------------------
+      -- Emit_Progress_End --
+      -----------------------
+
+      procedure Emit_Progress_End is
+         P : LSP.Messages.Progress_Params (LSP.Messages.Progress_End);
+      begin
+         P.End_Param.token := token;
+         Self.Server.On_Progress (P);
+      end Emit_Progress_End;
+
+      Index           : Natural := 1;
+      Total           : Natural := 0;
+      Last_Percent    : Natural := 0;
+      Current_Percent : Natural := 0;
+   begin
+      --  Collect the contexts and their sources: we do this so that we are
+      --  able to produce a progress bar covering the total number of files.
+
+      for J of Self.Contexts loop
```

Comment:
I would prefer to use a more explicit name here, like 'Context' (J is more used for integer-based indexes)

Reply:
Will do!

Reply:
done!